openthread: Move OpenThread implementation from net to modules

ArekBalysNordic opened the pull request:
Move OpenThread-related code from
zephyr/subsys/net/l2/openthread/openthread.c to
zephyr/modules/openthread/openthread_module.c.

The primary goal of this refactor is to enable the use of OpenThread as an independent module, without the necessity of Zephyr's networking layer.

This change is particularly beneficial for simple applications that have their own implementation of the IEEE802.15.4 driver and do not require a networking layer. These applications can now disable Zephyr's L2 and IEEE802.15.4 shim layers and directly use the OpenThread module, saving valuable kilobytes of memory.

In this approach if the CONFIG_NET_L2_OPENTHREAD
Kconfig option is set, Zephyr's L2 and IEEE802.15.4 layer will be used, and everything will function as before. The main difference is the Zephyr's L2 layer now uses the OpenThread module, no longer implementing it.

If the CONFIG_NET_L2_OPENTHREAD Kconfig option is
not set, the opentherad_init function will be called at system boot in the POST_KERNEL phase.

While most of the functions in include/net/openthread.h have been deprecated, they are still available for use to maintain backwards compatibility.
These functions now act as an interface to the new ones defined in modules/openthread.

adamkondraciuk committed:
[nrf fromlist] drivers: i2s: Add support for nRF TDM peripherals

Add a shim that allows using the nRF TDM
(Time division multiplexed audio interface) HAL by I2S Zephyr API.

Upstream PR #: 82144

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

wearyzen committed:
[nrf fromtree] modules: avoid fetching external repo in TF-M

What is changed?
 - Use the updated TF-M that is compatible with the Zephyr's
   latest Ethos-U driver repo.
 - Change the default behavior of TF-M builds to use Ethos driver
   locally fetched by Zephyr, using west update, instead of
   downloading it from external repo.

Why is this change required?
 - This is to be inline with Zephyr's rules to not fetch code from
   external repo.

Fixes #81656

Signed-off-by: Sudan Landge <sudan.landge@arm.com>
(cherry picked from commit 9b0623e63aa9df6ae64bf0df595e9a5d2f7871ba)

adamkondraciuk committed:
[nrf fromlist] tests: drivers: i2s: Align tests to TDM peripheral

Some of nRF54's has TDM peripheral instead of I2S.

Upstream PR #: 82144

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

FRASTM committed:
[nrf fromtree] west.yml : update modules/tee/tf-m/trusted-firmware-m

Updates the https://github.com/zephyrproject-rtos/trusted-firmware-m/
to fix the build error on stm32u5 due to gcc revision

Signed-off-by: Francois Ramu <francois.ramu@st.com>
(cherry picked from commit 676c3d2dbb3093216cb632f8756128504a85cad7)

JordanYates committed:
[nrf fromtree] modules: tf-m: fix build verbosity

Only set CMAKE_INSTALL_MESSAGE in a single location, remove the
unconditional overrides of the value in other locations.

This prevents dozens of rather pointless messages appearing in the build
log when CONFIG_TFM_BUILD_LOG_QUIET is set.
```
Installing: /home/jordan/code/workspace/build/nrf5340dk/...
```

Signed-off-by: Jordan Yates <jordan@embeint.com>
(cherry picked from commit b74b092f22ea29950f291104a84e3433f4efd7d6)

Vge0rge committed:
[nrf fromtree] manifest: Bring TF-M with nRF54L15 upstream support

Update west

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
(cherry picked from commit d8b87d62799febd0e8327f7937133eb5a609a015)

de-nordic committed:
[nrf fromtree] west.yml: Update TrustedFirmware-M with Flash Area cherry-picks

Update brings in flash_area_get_sector.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>
(cherry picked from commit 789e3f340e5019c18d973b25aa281b89afae1449)

gordonklaus committed:
[nrf noup] tests: bluetooth: tester:  Enable PSA RNG on nRF54H20

The PSA is a cryptographically secure random number generator.

It will be enabled by default, eventually,  For now, enable it manually.

Signed-off-by: Gordon Klaus <gordon.klaus@nordicsemi.no>

Vge0rge committed:
[nrf fromtree] manifest: Bring TF-M with initial nRF54L10 support

Brings some commits from the upstream TF-M which introduce
nRF54L10 ns support.

Signed-off-by: Georgios Vasilakis <georgios.vasilakis@nordicsemi.no>
(cherry picked from commit 27174731c769a3382e53810335cfe0ef5581959e)

tomi-font committed:
[nrf fromtree] manifest: tf-m: update to 2.1.2

Update TF-M to 2.1.2 from version 2.1.1.

Signed-off-by: Tomi Fontanilles <tomi.fontanilles@nordicsemi.no>
(cherry picked from commit dfd0dd04c1857d038107b78b325ae992016a66df)

ArekBalysNordic committed:
openthread: Move OpenThread implementation from net to modules

Move OpenThread-related code from
zephyr/subsys/net/l2/openthread/openthread.c to
zephyr/modules/openthread/platform/openthread.c.

The primary goal of this refactor is to enable the use
of OpenThread as an independent module, without the necessity
of Zephyr's networking layer.

This change is particularly beneficial for simple applications
that have their own implementation of the IEEE802.15.4 driver
and do not require a networking layer. These applications can
now disable Zephyr's L2 and IEEE802.15.4 shim layers and
directly use the OpenThread module, saving valuable kilobytes
of memory.

In this approach if the CONFIG_NET_L2_OPENTHREAD
Kconfig option is set, Zephyr's L2 and IEEE802.15.4 layer
will be used, and everything will function as before.
The main difference is the Zephyr's L2 layer now uses
the OpenThread module, no longer implementing it.

While most of the functions in include/net/openthread.h
have been deprecated, they are still available for use to
maintain backwards compatibility.

Signed-off-by: Arkadiusz Balys <arkadiusz.balys@nordicsemi.no>

jonathannilsen committed:
[nrf fromlist] modules: hal_nordic: bump regtool to 9.1.0

Upstream PR #: 88898

Changes for this package version:

* CTRLSEL is now set appropriately for the SPIS121 pins.
* DPPI.LINK configuration is now skipped when processing
  the 'dppic130' devicetree node.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

jonathannilsen committed:
Revert "[nrf noup] moduled: hal_nordic: require nrf-regtool"

This reverts commit d6cefa7cf72bfb85b7c185928044e880cf64e8f8.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

jonathannilsen committed:
[nrf noup] modules: hal_nordic: require nrf-regtool

Same as commit 6ec9d10 but with the REQUIRED keyword on its own
line to attempt to avoid a merge conflict when reverting/reapplying
this patch.

Signed-off-by: Jonathan Nilsen <jonathan.nilsen@nordicsemi.no>

SebastianBoe committed:
[nrf fromlist] west: runners: nrf_common: Temp hack for NRF54H20_IRON

Temp hack while waiting for NRF54H20_IRON support for Network in
nrfutil.

Upstream PR #: 88910

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

anangl committed:
[nrf fromlist] drivers: flash: nrf_qspi_nor: Prevent CPU hang when XIP is re-enabled

Add a simple non-XIP transaction before deactivating the QSPI after
a XIP transaction is performed. This prevents a CPU hang from occuring
when another XIP transaction is attempted after the QSPI is activated
again.

Upstream PR #: 88967

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

mstasiaknordic committed:
[nrf fromlist] tests: drivers: clock_control_api: Adjust to nRF54L09 and nRF54L20

nRF54L09 and nRF54L20 need double the time for startup of
their clocks.

Upstream PR #: 88956

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

adamkondraciuk committed:
Revert "[nrf fromlist] tests: drivers: i2s: Align tests to TDM peripheral"

This reverts commit 3c0b18489134cb3a3dc7882e6034d55a73604b3b.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

adamkondraciuk committed:
Revert "[nrf fromlist] drivers: i2s: Add support for nRF TDM peripherals"

This reverts commit f07fb31d8033a32a3d0d3f0ebcd3621001a5aad2.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

rlubos commented on modules/openthread/platform/openthread.h:
```diff
@@ -0,0 +1,170 @@
+ * @param aMessage The message to receive.
+ * @param aContext The context to pass to the callback.
+ */
+typedef void (*receiveCallback)(otMessage *aMessage, void *aContext);
```
nit: Function name and params should be in snake case given it's Zephyr's code. And perhaps it'd make sense to prefix the type with openthread, something like `openthread_receive_callback`.

rlubos commented on modules/openthread/platform/openthread.h:
```diff
@@ -0,0 +1,170 @@
+/**
+ * @brief Get OpenThread thread identification.
+ */
+k_tid_t openthread_platform_thread_id_get(void);
```
Need to include <zephyr/kernel/thread.h> to get access to the `k_tid_t` type

rlubos commented on modules/openthread/platform/openthread.h:
```diff
@@ -0,0 +1,170 @@
+ * @retval !NULL On success.
+ * @retval NULL  On failure.
+ */
+struct otInstance *openthread_platform_get_default_context(void);
```
This declaration seems redundant?

rlubos commented on modules/openthread/platform/openthread.h:
```diff
@@ -0,0 +1,170 @@
+ * @retval !NULL On success.
+ * @retval NULL  On failure.
+ */
+struct otInstance *openthread_platform_get_default_instance(void);
```
I just wonder if we need the `default` in the name, since we only have a single OT instance anyway, I'd just shorten it to `openthread_platform_get_instance()`

rlubos commented on modules/openthread/platform/openthread.c:
```diff
@@ -0,0 +1,418 @@
+	return k_mutex_lock(&openthread_lock, K_NO_WAIT);
+}
+
+void openthread_platform_mutex_unlock()
```
nit: missing `void` param here and in other functions.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -711,24 +353,63 @@ struct openthread_context *openthread_get_default_context(void)
+
+int openthread_start(struct openthread_context *ot_context)
+{
+	(void)ot_context;
```
`ARG_UNUSED(ot_context)`, same elsewhere.

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -287,8 +163,8 @@ static void ot_receive_handler(otMessage *aMessage, void *context)
 
-	NET_DBG("Injecting %s packet to Zephyr net stack",
-		PKT_IS_IPv4(pkt) ? "translated IPv4" : "Ip6");
+	NET_INFO("Injecting %s packet to Zephyr net stack",
```
Do we really want to bump the log level here, it'd cause massive log flood in default config I believe?

rlubos commented on modules/openthread/platform/openthread.h:
```diff
@@ -0,0 +1,170 @@
+	 */
+	void (*state_changed_cb)(otChangedFlags flags, void *context, void *user_data);
+
+	/** Context of the callback which will be passed to the state_changed_cb */
+	void *context;
+
+	/** User data if required */
+	void *user_data;
```
Now that I think of it, it seems to me that two void pointers for de facto user specified data seem a bit redundant. Perhaps we could just stick to user_data here, the callback would then interpret it however it wants. If it's a pointer to some specific context (like in case of L2), it could just include user data inside it.

And perhaps `state_changed_cb` could provide a poiner to the related `otInstance` instead of the context.

rlubos reviewed (commented):
Commented on some smaller stuff, but otherwise the separation seems pretty good. I wonder if we could have a overlay file to enable this standalone configuration in the upstream openthread shell sample (https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/openthread/shell) to ensure it's functional and does not bitrot over time.

Deprecations look fine, although a good migration guide note would be needed.

kkasperczyk-no commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -376,7 +217,7 @@ static enum net_verdict openthread_recv(struct net_if *iface, struct net_pkt *pk
 	if (pkt_list_peek(ot_context) == pkt) {
 		pkt_list_remove_last(ot_context);
-		NET_DBG("Got injected Ip6 packet, sending to upper layers");
+		NET_INFO("Got injected Ip6 packet, sending to upper layers");
```
As Robert mentioned there are log level changes here and there in this file. Is it intended?

kkasperczyk-no commented on modules/openthread/platform/openthread.c:
```diff
@@ -0,0 +1,418 @@
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(module_openthread, CONFIG_OPENTHREAD_PLATFORM_LOG_LEVEL);
```
All loggers from this directory are named like net_openthread_something. Maybe we should follow it and use `net_openthread`?

kkasperczyk-no commented on modules/openthread/platform/openthread.c:
```diff
@@ -0,0 +1,418 @@
+	openthread_platform_mutex_unlock();
+}
+
+static void ot_joiner_start_handler(otError error, void *context)
```
Mark context as unused argument.

kkasperczyk-no commented on modules/openthread/platform/openthread.c:
```diff
@@ -0,0 +1,418 @@
+	return len;
+}
+
+static void openthread_process(struct k_work *work)
```
Mark work as unused argument.

kkasperczyk-no commented on modules/openthread/platform/openthread.c:
```diff
@@ -0,0 +1,418 @@
+	}
+}
+
+static void ot_state_changed_handler(uint32_t flags, void *context)
```
Mark context as unused argument. I'm stopping here, please go through the rest of the functions and make sure that unused arguments are marked as unused.

mstasiaknordic committed:
[nrf fromlist] tests: drivers: clock_control_api: move startup time to Kconfig

Moved target dependent startup time from header file to
Kconfig option to allow adding new targets with new .conf
file rather that modyfing source of the test. Adjusted
startup time for nRF54L09 and nRF54L20.

Upstream PR #: 88956

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

mstasiaknordic committed:
Revert "[nrf fromlist] tests: drivers: clock_control_api: Adjust to nRF54L09 and nRF54L20"

This reverts commit 0afd70bfd6e65c619b1fdfd4f08cb063922c0b3e.

Signed-off-by: Michał Stasiak <michal.stasiak@nordicsemi.no>

ArekBalysNordic commented:
@kkasperczyk-no @rlubos I modified the approach with headers. Now it is only one global header `openthread.h` file. I reused as many functions as possible from there. Functions that use `openthread_context` have been deprecated, and I created new ones without this context alongside.
